<a href="https://colab.research.google.com/github/FachrizaFajri/UAS-DATA-MINING/blob/main/UAS_data_mining.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **MENDETEKSI STUNTING PADA BALITA MENGGUNAKAN ALGORITMA KNN**

nama anggota:


1.   Billy Restu Andrean
2.   Fourtyia Arteuary
3.   Fachriza Khairul Fajri
4.   Muhamad Dear Putra Naya
5.   Sahdan Pebriansah Arifin



Pada analisis project ini, akan dilakukan proses klasifikasi atau pembangunan model machine learning, yang lebih spesifiknya menggunakan algoritma KNN. Model ini akan dibangun dengan menggunakan data sampel yang berkaitan dengan stunting, khususnya umur dan tinggi badan seorang balita. Berdasarkan data tersebut, model diharapkan mampu mengklasifikasikan apakah seseorang terindikasi mengalami stunting atau tidak.

Dataset yang di gunakan adalah https://www.kaggle.com/datasets/rendiputra/stunting-balita-detection-121k-rows berdasarkan data umur dan tinggi badan, dapat dilakukan penentuan apakah seseorang didiagnosis mengalami stunting atau tidak

# IMPORT LIBRARY

In [ ]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score
from sklearn.metrics import classification_report, confusion_matrix



# IMPORT DATA

In [ ]:
df = pd.read_csv("/content/drive/MyDrive/latihan data mining/data_balita ieuuuuuuuuu.csv", delimiter=';')
df

Umur (bulan) Jenis Kelamin  Tinggi Badan (cm)       Status Gizi
0                  0     laki-laki          44.591973           stunted
1                  0     laki-laki          56.705203            tinggi
2                  0     laki-laki          46.863358            normal
3                  0     laki-laki          47.508026            normal
4                  0     laki-laki          42.743494  severely stunted
...              ...           ...                ...               ...
120994            60     perempuan         100.600000            normal
120995            60     perempuan          98.300000           stunted
120996            60     perempuan         121.300000            normal
120997            60     perempuan         112.200000            normal
120998            60     perempuan         109.800000            normal

[120999 rows x 4 columns]

In [ ]:
df.head()

Umur (bulan) Jenis Kelamin  Tinggi Badan (cm)       Status Gizi
0             0     laki-laki          44.591973           stunted
1             0     laki-laki          56.705203            tinggi
2             0     laki-laki          46.863358            normal
3             0     laki-laki          47.508026            normal
4             0     laki-laki          42.743494  severely stunted

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 120999 entries, 0 to 120998
Data columns (total 4 columns):
 #   Column             Non-Null Count   Dtype  
---  ------             --------------   -----  
 0   Umur (bulan)       120999 non-null  int64  
 1   Jenis Kelamin      120999 non-null  object 
 2   Tinggi Badan (cm)  120999 non-null  float64
 3   Status Gizi        120999 non-null  object 
dtypes: float64(1), int64(1), object(2)
memory usage: 3.7+ MB


In [ ]:
category_counts = df['Status Gizi'].value_counts()
category_counts

Status Gizi
normal              67755
severely stunted    19869
tinggi              19560
stunted             13815
Name: count, dtype: int64

# PRA PEMRORESAN DATA (ENCODING)

In [ ]:
# mengkodekan jenis kelamin
df['Jenis Kelamin'] = df['Jenis Kelamin'].map({'laki-laki': 0, 'perempuan': 1})

# mengkodekan status gizi
df['Status Gizi'] = df['Status Gizi'].map({'severely stunted': 0, 'stunted': 1, 'normal': 2, 'tinggi': 3})
df

Umur (bulan)  Jenis Kelamin  Tinggi Badan (cm)  Status Gizi
0                  0              0          44.591973            1
1                  0              0          56.705203            3
2                  0              0          46.863358            2
3                  0              0          47.508026            2
4                  0              0          42.743494            0
...              ...            ...                ...          ...
120994            60              1         100.600000            2
120995            60              1          98.300000            1
120996            60              1         121.300000            2
120997            60              1         112.200000            2
120998            60              1         109.800000            2

[120999 rows x 4 columns]

Tipe data pada kolom jenis kelamin sudah di ubah agar mempermudah proses,yang dimana angka 0 merepresentasikan sebagai laki laki dan 1 sebagai perempuan
dan juga merubah tipe data pada status gizi yaitu 0 berarti sangat terhambat 1 stunting 2  normal 3 tinggi


# PEMISAHAN DATA TRAIN DAN DATA TEST

In [ ]:
# memisahkan fitur dan target
X = df.iloc[:, 1:6]
X = df.drop('Status Gizi', axis=1)
y = df['Status Gizi']



In [ ]:
# membagi dataset menjadi data latih dan data uji
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
X

Umur (bulan)  Jenis Kelamin  Tinggi Badan (cm)
0                  0              0          44.591973
1                  0              0          56.705203
2                  0              0          46.863358
3                  0              0          47.508026
4                  0              0          42.743494
...              ...            ...                ...
120994            60              1         100.600000
120995            60              1          98.300000
120996            60              1         121.300000
120997            60              1         112.200000
120998            60              1         109.800000

[120999 rows x 3 columns]

dengan fungsi train_test_split() dari scikit-learn, kita bagi data menjadi:

80% untuk data latih, yang digunakan untuk melatih model

20% untuk data uji, yang digunakan untuk menguji performa model


Kita juga menggunakan parameter random_state=42 untuk memastikan hasil pembagian konsisten jika dijalankan ulang.


Setelah itu, kita inisialisasi model KNN menggunakan KNeighborsClassifier() dengan:


In [ ]:
# membuat objek KNN dengan k=3
knn = KNeighborsClassifier(algorithm='kd_tree', n_neighbors=1)

# melatih model KNN dengan data latih
knn.fit(X_train, y_train)


KNeighborsClassifier(algorithm='kd_tree', n_neighbors=1)

# EVALUASI MODEL

In [ ]:
# membuat prediksi dengan data uji
y_pred = knn.predict(X_test)

# menghitung akurasi model
accuracy = accuracy_score(y_test, y_pred)
print('Akurasi model KNN:', accuracy)

Akurasi model KNN: 0.9990495867768595


Angka ini, 0.9990, menunjukkan bahwa model KNN kita memiliki tingkat akurasi sekitar 99.9%. Ini adalah angka yang sangat tinggi, menunjukkan bahwa model kita mampu memprediksi kelas target dengan tingkat ketepatan yang luar biasa pada data yang belum pernah dilihat sebelumnya.

In [ ]:

print('confusion_matrix knn:')
confusion_matrix(y_test, y_pred)

confusion_matrix knn:


array([[ 4129,     1,     0,     0],
       [    6,  2782,     2,     0],
       [    0,     3, 13372,     7],
       [    0,     0,     4,  3894]])

In [ ]:
print(classification_report(y_test.values.reshape(-1, 1), y_pred))


              precision    recall  f1-score   support

           0       1.00      1.00      1.00      4130
           1       1.00      1.00      1.00      2790
           2       1.00      1.00      1.00     13382
           3       1.00      1.00      1.00      3898

    accuracy                           1.00     24200
   macro avg       1.00      1.00      1.00     24200
weighted avg       1.00      1.00      1.00     24200



Ini adalah hasil yang mengindikasikan bahwa model kita tidak hanya akurat secara keseluruhan, tetapi juga sangat andal dan robust dalam memprediksi setiap kategori data yang berbeda."

## UJI DATA BARU

In [ ]:
# Testing model
test = pd.DataFrame({'Umur (bulan)': [10], 'Jenis Kelamin': [1], 'Tinggi Badan (cm)': [86.2]})

pred = knn.predict(test)
print("Predicted Score = {}".format(np.array(pred)))

Predicted Score = [3]


Di sini, kami buat data baru yang merepresentasikan satu anak:
Usia 10 bulan
Jenis kelamin perempuan (yang dikodekan menjadi 1)
Tinggi badan 86.2 cm
Kita masukkan data ini dalam bentuk DataFrame, lalu prediksi status gizinya menggunakan knn.predict().

Hasil prediksi ditampilkan sebagai angka, misalnya "2", yang berarti "normal" sesuai dengan konversi label sebelumnya.


Kesimpulannya, metode KNN sangat mudah diimplementasikan dan cocok untuk klasifikasi sederhana seperti ini.
Meskipun sederhana, akurasinya cukup tinggi dan hasilnya bisa digunakan sebagai bahan pertimbangan dalam pengambilan keputusan.


source kode:https://www.kaggle.com/code/rendiputra/stunting-detection-with-knn-algorithm

source dataset:https://www.kaggle.com/code/rendiputra/stunting-detection-with-knn-algorithm